File: app/src/main/java/com/zack/enderplan/view/dialog/TypeMarkPatternPickerDialogFragment.java
package com.zack.enderplan.view.dialog;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.GridView;

import com.zack.enderplan.R;
import com.zack.enderplan.model.DataManager;
import com.zack.enderplan.model.bean.TypeMarkPattern;
import com.zack.enderplan.util.ResourceUtil;
import com.zack.enderplan.view.adapter.TypeMarkPatternGridAdapter;

import java.util.List;

import butterknife.BindView;

public class TypeMarkPatternPickerDialogFragment extends BaseDialogFragment {

    @BindView(R.id.grid_type_mark_pattern)
    GridView mTypeMarkPatternGrid;

    private static final String ARG_DEFAULT_PATTERN = "default_pattern";

    private TypeMarkPattern mTypeMarkPattern;
    private int mPosition = -1;
    private List<TypeMarkPattern> mTypeMarkPatternList;
    private OnTypeMarkPatternPickedListener mOnTypeMarkPatternPickedListener;

    public TypeMarkPatternPickerDialogFragment() {

    }

    public static TypeMarkPatternPickerDialogFragment newInstance(String defaultPattern) {
        TypeMarkPatternPickerDialogFragment fragment = new TypeMarkPatternPickerDialogFragment();
        Bundle args = new Bundle();
        args.putString(ARG_TITLE, ResourceUtil.getString(R.string.title_dialog_fragment_type_mark_pattern_picker));
        args.putString(ARG_NEG_BTN, ResourceUtil.getString(R.string.button_remove));
        args.putString(ARG_POS_BTN, ResourceUtil.getString(R.string.button_select));
        args.putString(ARG_DEFAULT_PATTERN, defaultPattern);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        String defaultPattern = null;
        Bundle args = getArguments();
        if (args != null) {
            defaultPattern = args.getString(ARG_DEFAULT_PATTERN);
        }

        mTypeMarkPatternList = DataManager.getInstance().getTypeMarkPatternList();

        mTypeMarkPattern = new TypeMarkPattern();

        if (defaultPattern != null) {
            for (int i = 0; i < mTypeMarkPatternList.size(); i++) {
                TypeMarkPattern typeMarkPattern = mTypeMarkPatternList.get(i);
                if (typeMarkPattern.getPatternFn().equals(defaultPattern)) {
                    mPosition = i;
                    mTypeMarkPattern.setPattern(typeMarkPattern.getPatternFn(), typeMarkPattern.getPatternName());
                    break;
                }
            }
        }
    }

    @Override
    public View onCreateContentView(LayoutInflater inflater, ViewGroup root) {
        return inflater.inflate(R.layout.dialog_fragment_type_mark_pattern_picker, root, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mTypeMarkPatternGrid.setAdapter(new TypeMarkPatternGridAdapter(mTypeMarkPatternList));

        if (mPosition != -1) {
            mTypeMarkPatternGrid.setItemChecked(mPosition, true);
        }

        mTypeMarkPatternGrid.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                mPosition = position;
                TypeMarkPattern typeMarkPattern = mTypeMarkPatternList.get(mPosition);
                mTypeMarkPattern.setPattern(typeMarkPattern.getPatternFn(), typeMarkPattern.getPatternName());
            }
        });
    }

    @Override
    public boolean onButtonClicked(int which) {
        switch (which) {
            case BTN_NEG:
                if (mOnTypeMarkPatternPickedListener != null) {
                    mOnTypeMarkPatternPickedListener.onTypeMarkPatternPicked(null);
                }
                break;
            case BTN_POS:
                if (mOnTypeMarkPatternPickedListener != null) {
                    mOnTypeMarkPatternPickedListener.onTypeMarkPatternPicked(mTypeMarkPattern.getPatternFn() == null ? null : mTypeMarkPattern);
                }
                break;
        }
        return true;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mOnTypeMarkPatternPickedListener = null;
    }

    public interface OnTypeMarkPatternPickedListener {
        void onTypeMarkPatternPicked(TypeMarkPattern typeMarkPattern);
    }

    public void setOnTypeMarkPatternPickedListener(OnTypeMarkPatternPickedListener listener) {
        mOnTypeMarkPatternPickedListener = listener;
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/dialog/TypeMarkPatternPickerDialogFragment.kt
package net.zackzhang.app.end.view.dialog

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import kotlinx.android.synthetic.main.dialog_fragment_type_mark_pattern_picker.*
import net.zackzhang.app.end.R
import net.zackzhang.app.end.model.DataManager
import net.zackzhang.app.end.model.bean.Type
import net.zackzhang.app.end.model.bean.TypeMarkPattern
import net.zackzhang.app.end.util.ResourceUtil
import net.zackzhang.app.end.view.adapter.TypeMarkPatternGridAdapter
import me.imzack.lib.basedialogfragment.BaseDialogFragment

class TypeMarkPatternPickerDialogFragment : BaseDialogFragment() {

    companion object {

        private const val ARG_DEFAULT_PATTERN = "default_pattern"

        fun newInstance(defaultPattern: String?): TypeMarkPatternPickerDialogFragment {
            val fragment = TypeMarkPatternPickerDialogFragment()
            val args = Bundle()
            putAddHorizontalMargins(args, false)
            args.putString(ARG_DEFAULT_PATTERN, defaultPattern)
            fragment.arguments = args
            return fragment
        }
    }

    private val defaultPattern by lazy { arguments!!.getString(ARG_DEFAULT_PATTERN) }
    private val typeMarkPatternList = DataManager.typeMarkPatternList
    private var position = -1
    private val typeMarkPattern by lazy { if (position == -1) TypeMarkPattern() else typeMarkPatternList[position].copy() }

    var typeMarkPatternPickedListener: ((typeMarkPattern: TypeMarkPattern?) -> Boolean)? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // 为 position 设置正确的初始值
        if (Type.hasMarkPattern(defaultPattern)) {
            // indexOfFirst 找到第一个出现符合要求的元素下标，如果没找到，返回 -1（实际上不会出现）
            position = typeMarkPatternList.asSequence().indexOfFirst { it.file == defaultPattern }
        }
    }

    override fun onCreateContentView(inflater: LayoutInflater, root: ViewGroup) =
            inflater.inflate(R.layout.dialog_fragment_type_mark_pattern_picker, root, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        titleText = ResourceUtil.getString(R.string.title_dialog_fragment_type_mark_pattern_picker)

        grid_type_mark_pattern.adapter = TypeMarkPatternGridAdapter(typeMarkPatternList)

        if (position != -1) {
            grid_type_mark_pattern.setItemChecked(position, true)
        }

        grid_type_mark_pattern.onItemClickListener = AdapterView.OnItemClickListener { _, _, pos, _ ->
            position = pos
            val (patternFn, patternName) = typeMarkPatternList[position]
            typeMarkPattern.file = patternFn
            typeMarkPattern.name = patternName
        }

        neutralButtonText = ResourceUtil.getString(R.string.button_remove)
        neutralButtonClickListener = { typeMarkPatternPickedListener?.invoke(null) != false }
        negativeButtonText = ResourceUtil.getString(android.R.string.cancel)
        negativeButtonClickListener = { true }
        positiveButtonText = ResourceUtil.getString(R.string.button_select)
        // takeUnless 判断调用它的变量是否符合要求，如果符合，返回此变量，否则返回 null
        positiveButtonClickListener = { typeMarkPatternPickedListener?.invoke(typeMarkPattern.takeUnless { it.file == null }) != false }
    }
}
